Fixing LuaLS problems, general cleanup and improvment around FactionTemplate

Logonz opened the pull request:
Fix a lot of the problems that arise from the language server to make the development of Questie be easier.

Also fixed a bit with FactionTemplate generation and overriding.

# What needs reviewing:
Mostly the code files, ignore everything that is .csv and that exists in the .types catalog as that is to help the rest of the codebase.

Logonz committed:
Ace Types

Logonz committed:
[gitignore] Add WoW-API to .gitignore

Logonz committed:
[types] Enhance quest and object definitions with Enum.QuestIcon and waypoints

- Added Enum.QuestIcon to various objective classes in Quest.t.lua for better icon management.
- Introduced waypoints field in Object.t.lua to track object locations.
- Defined a new Point class in Map.t.lua for coordinate representation.

Logonz committed:
[refactor] Improve error handling and cleanup in validation scripts

- Added assertions to ensure files are opened successfully in loadTOC functions across multiple validation scripts.
- Removed unnecessary function parameters in Compile calls for better clarity.
- Added diagnostic comments to suppress warnings for fields that are not required in minimal setups.

Logonz committed:
[types] Update type annotations for various modules

- Changed type annotations from `---@type` to `---@class` for QuestieDB, SeasonOfDiscovery, and ZoneDB in multiple files to improve clarity and consistency.
- Updated function parameters in QuestieProfiler and QuestieStream to use underscore for unused variables, enhancing code readability.
- Refactored options modules to ensure proper class annotations for QuestieOptions and its tabs.

Logonz committed:
[types] Update type annotations and improve code readability

- Added diagnostic comments to suppress warnings for unused fields and parameters in various modules, including QuestieDB, QuestieStream, and QuestieJourney.
- Refactored loops to use underscore for unused variables, enhancing clarity in QuestEventHandler and QuestieJourneyUtils.
- Introduced new type annotations for better documentation and understanding of the code structure.

Logonz committed:
[types] Update function parameters and type annotations in Questie

- Modified the RefreshConfig function to use underscores for unused parameters, improving code clarity.
- Updated the Colorize function's type annotation to allow for a nil value for the color parameter, enhancing flexibility in usage.

Logonz committed:
[fix] Improve NPC, object, item, and quest validation logic

- Updated validation functions in QuestieDBCompiler to handle cases where compiledData may be nil, preventing potential errors during data validation.
- Ensured that checks for questLevel and requiredLevel are only performed when compiledData is available, enhancing robustness of the validation process.

Logonz committed:
[types] Update QuestieDB with new field and improve objective handling

- Added a new field `tagInfoWasCached` to the quest object for better caching information.
- Enhanced the handling of reputation objectives by including an `Icon` field for improved data representation.
- Refactored loop variable usage for clarity in processing spell objectives.

Logonz committed:
[fix] Correct quest zone sorting keys in SeasonOfDiscovery

- Updated the zoneOrSort keys for several quests to use the correct sorting keys instead of zone IDs, ensuring proper categorization of quests.
- Adjusted quests: Anyone Can Cook, A Quick Grocery Run, Taste Testing, and Aeonas the Vindicated to reflect these changes.

Logonz committed:
[types] Update parameter annotations in QuestieCompat

- Modified the SetResizeBounds function to allow maxWidth and maxHeight parameters to be optional, improving flexibility in usage.
- Enhanced the GetContainerItemInfo function to return additional nil values for better consistency in return types.

Logonz committed:
[fix] Refactor date parsing in QuestieEvent

- Updated variable names for clarity in the date parsing logic within the Load function.
- Ensured that start and end dates are parsed correctly from strings to numbers, improving code readability and maintainability.

Logonz committed:
[types] Enhance IconFrame
- Improved waypoint addition logic to ensure proper checks against nil values, enhancing robustness.

Logonz committed:
[fix] Enhance quest objective validation in DebugFunctions

- Added checks to ensure the quest exists and has objective data before processing, improving error handling.
- Initialized additional fields in the quest cache to prevent nil values, enhancing data integrity.

Logonz committed:
[types] Enhance type annotations and improve code structure

- Added type annotations for various AceGUI components across multiple modules, including QuestieJourney, QuestieJourneyUtils, and QuestieSearchResults, to improve code clarity and maintainability.
- Refactored existing code to ensure consistent formatting and spacing, enhancing overall readability.
- Updated function parameters and added diagnostic comments to suppress warnings for unused fields, improving documentation and understanding of the code structure.

Logonz committed:
[refactor] Improve module structure and initialization

- Added private tables to ZoneDB, QuestieInit, and QuestieTooltips for better encapsulation.

Logonz committed:
[refactor] Standardize function parameter handling across modules

- Updated function signatures in QuestieOptionsGeneral, QuestieOptionsIcons, QuestieOptionsNameplate, and QuestieOptionsTracker to use underscores for unused parameters, improving code clarity and consistency.
- Enhanced readability of description strings by formatting them for better presentation.
- Refactored logic in various functions to ensure uniformity in handling parameters and improve overall maintainability.

Logonz committed:
[refactor] Simplify loop variable usage in QuestieStream

- Updated loop variable names in multiple functions within QuestieStream to use underscores for unused variables, enhancing code clarity and consistency.

Logonz committed:
[types] Enhance type annotations and improve quest handling

- Added new type annotations for various fields in QuestieFrame, QuestieFramePool, and QuestLogCache to improve code clarity and maintainability.
- Updated function signatures to reflect changes in parameter types, enhancing consistency across modules.
- Improved error handling in AvailableQuests by ensuring quest data is validated before processing, preventing potential nil reference issues.
- Refactored logic in MapIconTooltip to streamline tooltip display and improve readability.

Logonz committed:
[fix] Improve quest handling in TrackerMenu

- Added a nil check for quest_wow before accessing its name to prevent potential errors.
- Updated the addTomTomOptionForQuest function call to remove the unnecessary nil parameter, enhancing clarity and functionality.

Logonz committed:
[fix] Update quest event handling in EventHandler

- Modified the QUEST_TURNED_IN event registration to include nil parameters for improved compatibility with the QuestEventHandler.
- Simplified the instance check by removing the unused variable, enhancing code clarity and maintainability.

Logonz committed:
[fix] Add nil check for profile in MinimapIconFadeLogic

- Implemented a check to ensure the profile is not nil before proceeding with minimap icon logic, improving error handling and preventing potential crashes.
- Added an error message for better debugging when the profile is nil, enhancing overall robustness of the function.

Logonz committed:
[refactor] Update _DoYell function signature in QuestieComms

- Modified the _DoYell function to remove the questId parameter, improving clarity and consistency in function usage.
- Updated comments to reflect the change, enhancing code documentation.

Logonz committed:
[fix] Improve nil checks and error handling across quest modules

- Added nil checks for various quest-related objects in QuestEventHandler, QuestieQuest, AvailableQuests, and Link modules to prevent potential errors and crashes.
- Enhanced the return values in TrackerQuestTimers to provide more informative outputs, improving error handling.
- Updated comments and code structure for better clarity and maintainability.

Logonz committed:
[fix] Enhance keybind validation and tooltip handling

- Improved the IsBindTrue function to include better nil checks and return false when conditions are not met, enhancing error handling.
- Updated tooltip display logic in TrackerItemButton to remove unnecessary self references, streamlining the code and improving clarity.

Logonz committed:
[fix] Refactor tooltip display logic in MapIconTooltip

- Updated the QuestieLib:TextWrap function call to remove the unnecessary questId parameter, enhancing clarity.
- Modified the QuestXP function call to use dot notation for consistency.
- Streamlined the AddLine function call by removing an unused parameter, improving code readability.

Logonz committed:
[refactor] Simplify object count retrieval in TooltipHandler

- Replaced the table.getn function with the more concise # operator for retrieving the count of object names in the tooltip handler, enhancing code readability and consistency.

Logonz committed:
Merge commit '8ae2fac7117f3aff3f614d563440c61b0a1024b8' into fixing-languageserver-problems

Logonz committed:
[fix] Add diagnostic directive for missing fields in cata validation

- Introduced a diagnostic directive to suppress warnings for missing fields in the Questie database initialization, improving compatibility with language server features and enhancing code clarity.

Logonz committed:
[fix] Enhance NPC and object quest validation error handling

- Added checks to ensure correction files are successfully opened before writing, improving error handling and providing feedback if file operations fail.
- Streamlined the output of invalid quest starts and ends for both NPCs and objects, enhancing clarity in diagnostic messages.
- Updated return types in function documentation to reflect changes in output structure, improving code documentation.

Logonz committed:
[fix] Update return type in checkObjectQuestEnds function documentation

Logonz committed:
[fix] Update parameter name in AceGUITabGroup:SetTabs documentation so it is not confused with the type "table"

Logonz committed:
Small fix with missing colon

Logonz committed:
[fix] Update parameter names in AceGUI widget documentation for clarity

- Changed parameter names from generic "table" to more descriptive names like "statusTable", "groupList", and "list" in various AceGUI widget functions.
- Improved documentation for the SetType function in AceGUICheckBox to clarify the expected parameter type.
- Ensured consistency in documentation formatting across multiple widget files.

Logonz committed:
[fix] Add diagnostic directives to suppress unused variable warnings in the library HBD

Logonz committed:
Added skip for unused-local

Logonz committed:
[fix] Add diagnostic directive to suppress unused-local warning for _DoYell function

Logonz committed:
[cleanup] Remove outdated TODO comments in tracker initialization

Logonz committed:
[cleanup] Initialize QuestieOptions tabs as empty tables instead of using addon namespace

Logonz committed:
[db] Added _factionTemplateFixes.lua for faction template corrections
[maint] Updated faction template files to reference _factionTemplateFixes.lua
[tooling] Implemented downloadFactionTemplate.py for downloading CSV files
[tooling] Enhanced generateFactionTemplate.py to generate Lua files from CSV data

Logonz committed:
Added the FactionTemplate files.

Logonz committed:
[tooling] Added Copilot commit message generation instructions file

Logonz committed:
[ui] Updated schema location in embeds.xml for improved compatibility

Logonz committed:
[refactor] Removed protected access modifier from frame fields in AceGUIWidget, AceGUIContainer, and AceGUITreeGroup classes to prevent errors.

Logonz committed:
[refactor] Added nil check before canceling timer in ThreadLib.Thread function
[maint] Updated return type documentation for ThreadLib functions to TimerCallback

Logonz committed:
[maint] Commented out unused AddParagraph function in QuestieSearchResults.lua

Logonz committed:
[maint] Added diagnostic disable for unused-local in dump.lua
[maint] Added diagnostic disable-next-line for cast-local-type in profiler.lua
[fix] Asserted fileWriter existence before writing in profiler.lua

Logonz committed:
[maint] Updated ignore patterns in settings.json for improved file management
[maint] Cleaned up Lua diagnostics settings in settings.json
[maint] Removed unnecessary entries from workspace.library in settings.json

Logonz committed:
[maint] Added indentation style, size, and end of line settings to Lua format configuration

Logonz committed:
[db] Added ClassKeys class definition to QuestieDB
[maint] Clarified alias for Classes in DataGeneral.t.lua to fix references to "Classes" as string

Logonz committed:
[fix] Corrected player position retrieval in ProcessShownMinimapIcons function

Logonz committed:
[maint] Updated XML namespace declarations in lookupObjects.xml for improved compatibility

Logonz committed:
[maint] Cleaned up formatting in QuestieMenu.lua for consistency

Logonz committed:
[maint] Updated comments to clarify the exclusion of redundant parameters in QuestieTracker.lua

Logonz committed:
[maint] Commented out unused factionIDs variable in QuestieQuestFixes:Load function

Logonz committed:
Merge remote-tracking branch 'origin/master' into fixing-languageserver-problems

Logonz committed:
[fix] Simplified waypoint addition logic in OnClick function

Logonz committed:
Inline set a object if nothing is returned.

Logonz committed:
Remove the if case

Logonz committed:
[maint] Clarified return type documentation for GetNewObjectives function

Logonz committed:
[maint] Clarified parameter type for Colorize function

Logonz committed:
[maint] Clarified documentation regarding Enum.ItemQuality naming conventions
[fix] Adjusted item quality check to accommodate both Standard and Common values

Logonz committed:
[refactor] Simplified OnLeave script by removing unnecessary self parameter.

BreakBB commented on Database/Corrections/sodItemFixes.lua:
```diff
@@ -1,4 +1,4 @@
----@type SeasonOfDiscovery
+---@class SeasonOfDiscovery
```
I still think it's wrong to have multiple `class` declarations of the same class. I do get why you changed it this way though.

BreakBB commented on Modules/Journey/QuestieSearchResults.lua:
```diff
@@ -100,6 +106,7 @@ local function CreateShowHideButton(id)
         button:SetCallback("OnClick", function(self) self:RemoveFromMap(self) end)
```
```suggestion
        button:SetCallback("OnClick", function(self) self:RemoveFromMap() end)
```

BreakBB commented on Modules/Journey/QuestieSearchResults.lua:
```diff
@@ -128,58 +136,62 @@ local function CreateShowHideButton(id)
 
 local function rec(theTable, ret, indent)
-    ret = ret..stringrep('    ', indent)..'{\n'
+    ret = ret .. stringrep('    ', indent) .. '{\n'
```
Do you have a formatter tool for this?

BreakBB reviewed (commented):
Most of the stuff is nitpicking and questions. Looks good to me overall 👍🏻 

I didn't look into LuaLS for a few month. Is it possible to have a check in the CI?

BreakBB committed:
Streamline luassert.spy chained calls

BreakBB committed:
Streamline luassert matcher calls

Logonz committed:
Clarified parameter type for questIdsToCheck in CheckForChanges function
Clarified return type documentation for CheckForChanges function

Logonz committed:
Merge branch 'fixing-languageserver-problems' of https://github.com/Questie/Questie into fixing-languageserver-problems

Logonz committed:
Merge remote-tracking branch 'origin/master' into fixing-languageserver-problems

Logonz committed:
This variable was incorrectly named

Logonz committed:
Removing unused and undefined fields.

Logonz committed:
Small changes to settings.json

Logonz committed:
Some ignores

Logonz committed:
Fixed with imports and definitions of modules.

Logonz committed:
Removed unused params and variables

Logonz committed:
Cleaned up unused variables

Logonz committed:
Fixed luarc settings and globals logic

Logonz committed:
Globals fixes.

Logonz committed:
Moved globals next to eachother.

Logonz committed:
Tracker Typing

Logonz committed:
Journey Ace typing

Logonz committed:
Set this as type otherwise the LSP is confused.

Logonz committed:
Icon and ObjectiveData typing

Logonz committed:
Ignores and typing

Logonz committed:
Formating and small fixes

Logonz committed:
Nil fixes

Logonz committed:
Criteria ID

Logonz committed:
Typing

Logonz committed:
Merge remote-tracking branch 'origin/master' into fixing-languageserver-problems

Logonz committed:
Merge commit '392bc97e851d95e9381b0881474d21b2973824f1' into fixing-languageserver-problems

Logonz committed:
Unused local

Logonz committed:
With the change to use the WoW-UI folder we can skip using Ketho.wow-api as it is already based on it.

Logonz committed:
These are not needed anymore.

Logonz committed:
We type the Questie Gametooltip widget

Logonz committed:
We can just work around this problem by checking.

Logonz committed:
Changed deprecated to "Information"

Exclude some of the libraries we handle types in other ways.

We need to disable this warning as the function does not exist in Era, and our WoW-API is generated from Era so we need to ignore it here.

Logonz committed:
unused local

Logonz committed:
Make LuaLS understand this is a global

Logonz committed:
Deprecated globals

Logonz committed:
Remove basically all lua settings and change CI to use LuaLS for diagnostics instead.

Logonz committed:
We have to download the WoW-API first...

Logonz committed:
[tooling] Updated Lua Language Server Check command to include metapath and logpath options.

Logonz committed:
Lets remove this commit message instruction

Logonz committed:
Simplified WoW-API fetch and moved the LuaLS to /tmp as well

Logonz committed:
Merge commit 'edf7c13b39e0348164bf4a49833fc04d42195e8f' into fixing-languageserver-problems
